### PR TITLE
Add Python 3.10 support (close #254)

### LINKED ISSUE #254: Add Python 3.10 support
Currently, Python 3.10 is not supported. Is it on the roadmap?

For what I know, `contracts` library may be an issue because it hasn't been updated in years and now some warnings turn into errors.

### PR DESCRIPTION
**Pull request by matus-tomlein**
This PR addresses issue #254 and adds support for Python 3.10 by dropping the pycontracts library (not updated for some time and incompatible with 3.10 at this time) and adding some other features that require support for Python 2 to be dropped.

The full list of changes is:

* Remove pycontracts library that wasn't compatible with Python 3.10
* Add Python type hints to all function arguments and return values
* Remove support for Python 2 due to the type hints
* Add custom contracts (e.g., non_empty_string) evaluated at the start of functions
* Run through Flake8 linter and correct style
* Add Dockerfile for running tests under all supported Python versions

Note:

* Python type hints are not enforced during runtime (they can be checked using `mypy` static type checking or through IDEs such as PyCharm) so some types of issues that previously raised errors no longer do so
   * in particular, this is the case for instance type matching, I didn't implement contracts for checking instance types – I don't think this is a standard behaviour in Python and IDEs should help with these types of errors using the type hints
   * I tried to add custom contracts for non-instance-type checks such non-empty strings, values larger than 0, and non-empty lists
* There is a potentially breaking change because we no longer raise `ContractNotRespected` error (from pycontracts) when contracts are violated but the standard `ValueError`

### CODE REVIEW

**Comment by paulboocock** [comment]
I'm also assigning @adatzer to review this. I'll hopefully get to it too shortly but I'd love her input on this too given she has been the most recent maintainer on this tracker.

**Review by paulboocock — COMMENTED** [review]
This is looking good to me. I'll let @adatzer have the final say on approving this PR though.

Just a note that when preparing the 0.10.0 release, we should add a table to the README to make it clear what versions of the tracker support what versions of python (i.e. Python 2.7 is 0.9.x only, Python 3+ is 0.10.x+)

File: snowplow_tracker/emitters.py
Comment (adatzer) [inline snowplow_tracker/emitters.py:46]: nit - would it make sense for these types to be defined in `snowplow_tracker.typing` as well?

File: run-tests.sh
Comment (matus-tomlein) [reply run-tests.sh]: You're right, I didn't there was 3.10.1 already, updated!

File: snowplow_tracker/emitters.py
Comment (matus-tomlein) [reply snowplow_tracker/emitters.py]: True, we don't need that if anymore, removed it.

File: snowplow_tracker/typing.py
Comment (matus-tomlein) [reply snowplow_tracker/typing.py:54]: Yes, the ellipsis can have more uses (in slices), but they are commonly used instead of `pass`. It just looks a bit nicer.

File: .github/workflows/ci.yml
Comment (matus-tomlein) [reply .github/workflows/ci.yml]: Yes, added 3.10 here as well. Just had to make it string, otherwise Github actions understood it as version 3.1.

File: snowplow_tracker/tracker.py
Comment (matus-tomlein) [reply snowplow_tracker/tracker.py]: Good catch! Changed the tuple to `Tuple[str, Any]` and moved it to typing.py since it's a bit too long.

File: snowplow_tracker/tracker.py
Comment (matus-tomlein) [reply snowplow_tracker/tracker.py]: I think you're right, having all the types in typing.py does clean up the code a bit. I tried to move everything over there.

File: snowplow_tracker/tracker.py
Comment (matus-tomlein) [reply snowplow_tracker/tracker.py:402]: The problem here is that the check for form types is performed after making the string lowercase – I don't know how to do it using Python type hints, I can't use lambdas or make the type check case insensitive...

File: snowplow_tracker/tracker.py
Comment (matus-tomlein) [reply snowplow_tracker/tracker.py]: Thanks, changed!

File: snowplow_tracker/tracker.py
Comment (matus-tomlein) [reply snowplow_tracker/tracker.py:118]: Yes, newer versions of Python can actually directly use the class (e.g., `Tracker`) but this is not supported in 3.7 or older so I had to use the string alternative.

File: snowplow_tracker/tracker.py
Comment (matus-tomlein) [reply snowplow_tracker/tracker.py]: Great idea! I moved the check to constraints.py. It doesn't really belong in the Tracker class, much better now. Hopefully nobody relies on that Tracker function in their code and not a breaking change :)

File: snowplow_tracker/tracker.py
Comment (matus-tomlein) [reply snowplow_tracker/tracker.py:530]: This is actually the same – when you use the float type hint, [int is implicitly accepted too.](https://stackoverflow.com/questions/50928592/mypy-type-hint-unionfloat-int-is-there-a-number-type)

File: snowplow_tracker/tracker.py
Comment (matus-tomlein) [reply snowplow_tracker/tracker.py:104]: This is actually the same – when you use the float type hint, [int is implicitly accepted too.](https://stackoverflow.com/questions/50928592/mypy-type-hint-unionfloat-int-is-there-a-number-type)

File: snowplow_tracker/tracker.py
Comment (matus-tomlein) [reply snowplow_tracker/tracker.py:582]: This is actually the same – when you use the float type hint, [int is implicitly accepted too.](https://stackoverflow.com/questions/50928592/mypy-type-hint-unionfloat-int-is-there-a-number-type)

File: snowplow_tracker/tracker.py
Comment (matus-tomlein) [reply snowplow_tracker/tracker.py:688]: This is actually the same – when you use the float type hint, [int is implicitly accepted too.](https://stackoverflow.com/questions/50928592/mypy-type-hint-unionfloat-int-is-there-a-number-type)

File: snowplow_tracker/tracker.py
Comment (matus-tomlein) [reply snowplow_tracker/tracker.py]: True, thanks! Also moved the type hint to typing.py to clean up a bit.

File: snowplow_tracker/tracker.py
Comment (matus-tomlein) [reply snowplow_tracker/tracker.py:302]: This is actually the same – when you use the float type hint, [int is implicitly accepted too.](https://stackoverflow.com/questions/50928592/mypy-type-hint-unionfloat-int-is-there-a-number-type)

File: snowplow_tracker/tracker.py
Comment (matus-tomlein) [reply snowplow_tracker/tracker.py:352]: This is actually the same – when you use the float type hint, [int is implicitly accepted too.](https://stackoverflow.com/questions/50928592/mypy-type-hint-unionfloat-int-is-there-a-number-type)

File: snowplow_tracker/test/unit/test_tracker.py
Comment (matus-tomlein) [reply snowplow_tracker/test/unit/test_tracker.py]: This is a really good catch! It actually means that the `ContractsDisabled` was useless since all the tests passed without it doing anything. I added the function call and removed `ContractsDisabled` from it where it wasn't necessary anymore. Also added a couple of tests that make use of it.

File: snowplow_tracker/test/unit/test_tracker.py
Comment (matus-tomlein) [reply snowplow_tracker/test/unit/test_tracker.py]: Commented above, thanks for this!

File: snowplow_tracker/subject.py
Comment (matus-tomlein) [reply snowplow_tracker/subject.py]: Yes, moved to typing.py.

File: Dockerfile
Comment (matus-tomlein) [reply Dockerfile]: Yes, updated to `3.10.1`.

File: snowplow_tracker/emitters.py
Comment (matus-tomlein) [reply snowplow_tracker/emitters.py]: Right, moved all of these types to typing.py.

**Comment by matus-tomlein** [comment]
Thanks a lot @adatzer for the thorough review! Those were really useful comments and also uncovered a few important problems.

Not sure about removing the `:type` pycontracts comments – they still have some value in documenting the types although now the type documentation is duplicated in the type hints too... Do you know if they are used when generating any API docs?

File: snowplow_tracker/tracker.py
Comment (adatzer) [reply snowplow_tracker/tracker.py:104]: Sorry for missing the float type hint! Thank you for clarifying!

File: snowplow_tracker/tracker.py
Comment (adatzer) [reply snowplow_tracker/tracker.py:402]: You are right, i missed this point and i wasn't able to find a way through Python type hints either.

**Review by adatzer — APPROVED** [review]
Really happy to approve from my side! 
Concerning the pycontracts' comments, they are not used to generate API docs as far as i know. They do indeed add some value, so no objection from me for them to stay.

**Pull request merged**
